File: 101-symmetric-tree/solution.py
from typing import Optional
from collections import deque


class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def isSymmetric(self, root: Optional[TreeNode]) -> bool:
    if root is None:
      return True

    left_deque: deque[Optional[TreeNode]] = deque([root.left])
    right_deque: deque[Optional[TreeNode]] = deque([root.right])
    while left_deque and right_deque:
      left, right = left_deque.popleft(), right_deque.popleft()
      if (left is None and right is not None) or (left is not None and right is None) or (left is not None and right is not None and left.val != right.val):
        return False

      if left is not None:
        left_deque.extend([left.left, left.right])
      if right is not None:
        right_deque.extend([right.right, right.left])

    return not left_deque and not right_deque

File: 101-symmetric-tree/solution.cpp
#include <deque>

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  bool isSymmetric(TreeNode *root)
  {
    if (root == nullptr)
    {
      return true;
    }

    std::deque<const TreeNode *> left_deque{root->left}, right_deque{root->right};
    while (!left_deque.empty() && !right_deque.empty())
    {
      const auto *left = left_deque.front();
      const auto *right = right_deque.front();
      left_deque.pop_front();
      right_deque.pop_front();
      if ((left == nullptr && right != nullptr) || (left != nullptr && right == nullptr) || (left != nullptr && right != nullptr && left->val != right->val))
      {
        return false;
      }

      if (left != nullptr)
      {
        left_deque.emplace_back(left->left);
        left_deque.emplace_back(left->right);
      }
      if (right != nullptr)
      {
        right_deque.emplace_back(right->right);
        right_deque.emplace_back(right->left);
      }
    }

    return left_deque.empty() && right_deque.empty();
  }
};